### PR TITLE
Initial implementation of package:checks

### PR DESCRIPTION
**Pull request by natebosch**
Implements the core mechanics for defining and using expectations, along
with some basic expectations for core types. Many more expectation
extensions will be added, including `deepEquals` which is already called
out in the docs as necessary to replace the behavior of `equals` from
Matcher.

High level design:

A `Check` is the target for extension methods. It may hold a value (if
it came from `checkThat`) or it may be a placeholder for a value that
could be checked (if it came from `describe`). In test code only the
extensions are visible, to write an expectation a second import exposes
a `context` field on the check which gives access to defining the
expectations.

Each expectation is synchronous or asynchronous, and may be a simple
clause on the value, or may extract a property from the value for
further checking. These are `expect/Async`, and `nest/Async`
respectively.

`expect` and `expectAsync` will add a clause to the list of clause:

```
Expected: a value that:
  some clause from `expect` or `expectAsync`
  another clause from `expect` or `expectAsync`
```

`nest` and `nestAsync` will create a nested context which can have it's
own clauses

```
Expected: a value that:
  has some property extracted by a `nest` or `nestAsync` callback that:
    some clause from an expectation on the nested Check
```

When a failure happens, the clause that failed is replace by the
"Actual" description from the `Rejection`

```
Expected: a value that:
  some clause that will pass
  some clause that will fail
Actual: a value that:
  some clause that will pass
  Actual: The description of the failing value
```

Expectation extensions are written as small as possible, with the
intention that they can be put together in interesting ways. For
instance instead of a `hasLength(int length)` expectation, the `length`
is extracted and a `.equals` expectation can be used on the extracted
value.

```dart
// BAD
checkThat(someList).hasLength(10);
// GOOD
checkThat(someList).length.equals(10);
```

Whenever possible, a name that matches an API on the class is used to
correspond to exactly that behavior. For instance
`Check<Iterable>.contains` use `Iterable.contains` in the
implementation, while `Check<Iterable>.containsThat` takes a callback to
match with the Checks library instead.

### CODE REVIEW

File: pkgs/checks/lib/src/extensions/math.dart [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  }
+
+  void isLessThan(num other) {
+    context.expect(() => ['is less than ${literal(other)}'], (actual) {
+      if (actual < other) return null;
+      return Rejection(
+          actual: literal(actual),
+          which: ['Is not less than ${literal(other)}']);
+    });
+  }
```
Comment (lrhn) [inline pkgs/checks/lib/src/extensions/math.dart:20]: (Probably also want `.isFinite`, `.isNaN` and `.abs()` (as `=> has((n) => n.abs(), '.abs()')`).
Maybe even the `floor`/`ceil`/`round` functions.

**Review by lrhn — APPROVED** [review]

File: pkgs/checks/lib/src/extensions/core.dart [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+          which: ['is a value that: ', ...indent(describe(condition))]);
+    });
+  }
+}
+
+extension BoolChecks on Check<bool> {
+  void isTrue() {
+    context.expect(() => ['is true'], (actual) {
+      if (actual) return null;
+      return Rejection(actual: literal(actual));
```
[the remark this answers is not in the document]
Comment (natebosch) [reply pkgs/checks/lib/src/extensions/core.dart]: I do like one-liners... I used to like them too much and had to dial it back 😄

File: pkgs/checks/lib/src/extensions/core.dart [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    });
+  }
+}
+
+extension EqualityChecks<T> on Check<T> {
+  /// Expects that the value is equal to [other] according to [operator ==].
+  void equals(T other) {
+    context.expect(() => ['equals ${literal(other)}'], (actual) {
+      if (actual == other) return null;
+      return Rejection(actual: literal(actual), which: ['is not equal']);
```
[the remark this answers is not in the document]
Comment (lrhn) [reply pkgs/checks/lib/src/extensions/core.dart]: I don't think adding the `==` clause is necessary. That's the default meaning of "equal" in Dart.

I'd use "are" in either case.
```
Expected: a in that:
  equals <2>
Actual: <1>
Which: are not equal
```
That states that "1" and "2" are not equal (the two previously following objects), and the the `int` should have been equal to the former, and by transitivity, is therefore not equal to the latter.

Singular "is equal" must be transitive (it takes two things to be equal, so the singular subject is not enough). The plural "are equal" can refer to multiple subjects.

(Also, "a int" might need just a little cleverness to change "a" to "an".)

File: pkgs/checks/lib/src/extensions/core.dart [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    }, atSameLevel: true);
+  }
+}
+
+extension HasField<T> on Check<T> {
+  /// Extracts a property of the value for further expectations.
+  ///
+  /// Sets up a clause that the value "has [name] that:" followed by any
+  /// expectations applied to the returned [Check].
+  Check<R> has<R>(R Function(T) extract, String name) {
```
[the remark this answers is not in the document]
Comment (lrhn) [reply pkgs/checks/lib/src/extensions/core.dart:23]: Why won't autocomplete work in the second position?

File: pkgs/checks/lib/src/extensions/async.dart [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+extension ChainAsync<T> on Future<Check<T>> {
+  /// Checks the expectations in [condition] against the result of this Future.
+  ///
+  /// Extensions written on [Check] cannot be invoked on [Future<Check>]. This
+  /// method allows adding expectations for the value without awaiting it.
+  ///
+  /// ```dart
+  /// await checkThat(someFuture).completes().that((r) => r.equals('expected'));
```
[the remark this answers is not in the document]
Comment (lrhn) [reply pkgs/checks/lib/src/extensions/async.dart:136]: Nitpickinly, "completes" in the language specification covers both evaluating to a value and throwing ("completes by throwing error *e* and stack trace *s*").

We could have

* `Check<Result<T>> completes()`
* `Future<Check<T>> get value`
* `Future<Check<E>> throws<E>()`

to cover all the cases. The `Result` type belongs to `package:async`, but we already have `StreamQueue` specific checks too. (Should those be in a helper library of `package:async` instead? I don't like adding a dependency on another package just to use `package:checks`.)
If you don't like `value`, could it be `completesWith` (short for `completesWithValue`)?

You can do:
```dart
  await checkThat(someFuture).value.then((c) => c.equals(42));
```

If we want to allow `await checkThat(someFuiture).value.equals(42)`, we can also define all the matches on `Check<T>` also on `Future<Check<T>>`, working "below the future" .
So
```dart
extension <T> on Future<Check<T>> {
  Future<void> equals(Object? value) => this.then((c) => c.equals(value));
  Future<Check<R>> that<R>(R check(T value)) => this.then((c) => c.that(check));
  // etc.
}
```
That's a lot of extra declarations just to allow you to directly work on the value below the `Future`
And you can already do that using `then` on the future anyway.


The `.value` getter/`completes()` function converts a `Check<Future<T>>` to a `Future<Check<T>>`, which swaps the monads. That's actually quite an interesting operation. Do we want the opposite, taking a `Future<Check<T>>` and getting a `Check<Future<T>>`? It's probably not particularly useful, since there aren't a lot of things you can do on a future other than wait for it to complete.

File: pkgs/checks/lib/src/extensions/math.dart [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import 'package:checks/context.dart';
+
+extension NumChecks on Check<num> {
+  void isGreaterThan(num other) {
```
[the remark this answers is not in the document]
Comment (lrhn) [reply pkgs/checks/lib/src/extensions/math.dart:4]: If you need two `<` and `>` operations, we should give you an `.inRange(a, b)` instead. It's cleaner.

You can still do
```dart
checkThat(Box(actual)).that((v) {
  v > 0;
  v < 10;
});
```
That's one of the reasons for having `that`, as I understand it.

We can have *both*: Use `<`, `>`, `<=` and `>=` where they work, use `.greaterThan`, `.greaterThanEqual`, `lessThen` and `lessThanEqual` where they shorter versions don't work. Use `inRange` when you need to use two checks (on integers, it can be inclusive, on doubles, we might want to have both open and closed ranges).

File: pkgs/checks/lib/src/extensions/core.dart [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+extension StringChecks on Check<String> {
+  /// Expects that the value contains [pattern] according to [String.contains];
+  void contains(Pattern pattern) {
+    context.expect(() => ['contains $pattern'], (actual) {
+      if (actual.contains(pattern)) return null;
+      return Rejection(
+          actual: literal(actual), which: ['Does not contain $pattern']);
+    });
+  }
```
[the remark this answers is not in the document]
Comment (lrhn) [reply pkgs/checks/lib/src/extensions/core.dart:122]: I know, ignore all my ideas for extension (for now!)

File: pkgs/checks/lib/src/extensions/math.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (natebosch) [reply pkgs/checks/lib/src/extensions/math.dart:4]: > we should give you an `.inRange(a, b)` instead. It's cleaner.

I think I disagree, it's either too verbose to too cryptic to specify the open/closed ranges. I think seeing the checks for each boundary is going to be more in readable. I think this becomes especially true if we switch to `>` and `>=` instead of `isGreatherThan` and `isGreaterOrEqualTo`.

> You can still do
> 
> ```dart
> checkThat(Box(actual)).that((v) {
>   v > 0;
>   v < 10;
> });
> ```

Hmm, yeah I'm not sure why I was so fixated on cascades.

I think this one could be controversial, but it's one that I suspect might get more comfortable after usage so I might give it a try in these early versions and see how it goes.

File: pkgs/checks/lib/src/extensions/iterable.dart [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      if (actual.contains(element)) return null;
+      return Rejection(
+          actual: literal(actual),
+          which: ['does not contain ${literal(element)}']);
+    });
+  }
+
+  /// Expects that the iterable contains any element such that
+  /// [elementCondition] is satisfied.
+  void containsThat(void Function(Check<T>) elementCondition) {
```
[the remark this answers is not in the document]
Comment (lrhn) [reply pkgs/checks/lib/src/extensions/iterable.dart:43]: I tried doing something like that in my own primitive prototype. I didn't get it working. The problem was that the `Check<R>` yo get from `Check<Iterable<R>>.any` is not like the other checks, and when you then want to mix that with futures, a `Check<Future<Iterable<R>>` is not the same as a `Iterable<Future<Check<R>>>`. The monads do not commute :frowning:.
(At least I didn't get it working, maybe it's just a matter of adding more complexity!)

File: pkgs/checks/lib/src/extensions/math.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply pkgs/checks/lib/src/extensions/math.dart:4]: I'm sure someone will define `.inRange` if we don't. Since it's all extension methods, anyone can add them, and they can do whichever variant they prefer - open, closed, half-open in either direction ... heck:
```dart
extension RangeCheckNum on Check<num> {
  void inRange({num? min, num? max, num? inf, num? sup}) {
    if (min != null) this >= min;
    if (max != null) this <= end;
    if (inf != null) this > inf;
    if (sup != null) this < sup;
  }
}

extension RangeCheckComparable<T extends Comparable<T>> on Check<T> {
  Check<int> compareTo(T other) => that((self) => self.compareTo(other));

  void inRange({T? min, T? max, T? inf, T? sup}) {
    if (min != null) this.compareTo(min) >= 0;
    if (max != null) this.compareTo(max) <= 0;
    if (inf != null) this.compareTo(inf) > 0;
    if (sup != null) this.compareTo(sup) < 0;
  }
}
```

and then you can just do `checkThat(string).length.inRange(min:0, max: 7)`.

File: pkgs/checks/lib/src/extensions/iterable.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (natebosch) [reply pkgs/checks/lib/src/extensions/iterable.dart]: I did rename `containsThat` to `any()`. I'm split about 50/50 on which name I like better.

I was unsuccessful finding a pattern to make `.any.someCheck()` work.

File: pkgs/checks/lib/src/extensions/core.dart [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+      if (softCheck(actual, condition) != null) return null;
+      return Rejection(
+          actual: literal(actual),
+          which: ['is a value that: ', ...indent(describe(condition))]);
+    });
+  }
+}
+
+extension BoolChecks on Check<bool> {
+  void isTrue() {
```
Comment (scheglov) [inline pkgs/checks/lib/src/extensions/core.dart:67]: Why is this a method, and not a getter?
It will never have formal parameters, or type parameters.
So, why to force the user to write `()`?

File: pkgs/checks/lib/src/extensions/core.dart [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+  /// way that would conflict.
+  ///
+  /// ```
+  /// checkThat(something)
+  ///   ..has((s) => s.foo, 'foo').equals(expectedFoo)
+  ///   ..has((s) => s.bar, 'bar').that((b) => b
+  ///     ..isLessThan(10)
+  ///     ..isGreaterThan(0));
+  /// ```
+  R that<R>(R Function(Check<T>) condition) => condition(this);
```
Comment (scheglov) [inline pkgs/checks/lib/src/extensions/core.dart:48]: Have you considered naming this method `which` instead?

It might be not quite right in English though, if so, I apologize :-)

My motivations are mostly aesthetic, seeing `checkThat() ... that()` seems somehow wrong to me (in Russian "that" in `checkThat()` "проверить что" and "that" later "который" have absolutely different meaning). And `that` looks close to `then`.

Comment (lrhn) [reply pkgs/checks/lib/src/extensions/core.dart:48]: Could get behind `which` too, as a another non-native English speaker.

I know that `that` is restrictive and `which` is descriptive in prose, so `that` is probably "more correct".

Reading it with a predicate
```dart
  checkThat(o).has((o) => o.foo).that(isTrue)
  checkThat(o).has((o) => o.foo).which(isTrue)
```
I can't say one reads strictly better than the other, but I kind-of *like* reading `which` more.
Maybe it's just the repetition of `that`.

Also, in
```dart
checkThat(Box(actual)).that((v) {
  v > 1;
  v < 4;
});
```
the `that` would work better as `where`, because it's not checking a single predicate. I think `which` is slightly better than `that`, but we are firmly in "personal preference" land with that.

And while `that` is close to `then`, `which` is equally close to `while` or `where`.

TL;DR: Slight preference for `which`.

Non-suggestion: `st`! `checkThat(o).st((v) ...)`. :wink:

File: pkgs/checks/lib/src/extensions/core.dart (hunk 9 above)
Comment (lrhn) [reply pkgs/checks/lib/src/extensions/core.dart:67]: Predicates are generally written like boolean checks (`contains`, `isNull`) on the expect object.

They probably should go all-in and be named like boolean getters or boolean functions would, and be getters when there are no arguments.

(Do we want an argument, like the `description` of an `expect` call?)
